### PR TITLE
🧑‍🍳 Add precompute batch size argument in `DPOTrainer` for reference model

### LINKED ISSUE #2421: Adding precompute batch size argument in DPOTrainer for reference model
### Feature request

Proposing adding a new configuration parameter `precompute_ref_batch_size` to allow users to specify a different (likely larger) batch size specifically for the reference model precomputation phase. This would:

1. Speed up the precomputation phase by processing more examples per batch
2. Make better use of available GPU memory since no gradients need to be stored
3. Maintain backward compatibility by defaulting to current behavior if not specified

The change would affect:
- DPOConfig: Add new optional parameter `precompute_ref_batch_size`
- `get_train_dataloader()` and `get_eval_dataloader()`: Use the new batch size when precomputing reference probabilities

This is particularly useful for large-scale DPO training where the precomputation phase can be a significant bottleneck.

### Motivation

Currently when `precompute_ref_log_probs=True` the `DPOTrainer` class uses the `per_device_train_batch_size` or `per_device_eval_batch_size` respectively during training or evaluation for generating the logprobs.
But for efficiency this batch size can be more than the used for training one since this step does not require gradient computation and storage in memory. Adding a configurable `precompute_ref_batch_size` parameter would allow users to optimize this preprocessing step by using larger batch sizes while maintaining memory efficiency.

### Your contribution

Yes, I can help by submitting a PR following the contribution guidelines.

### PR DESCRIPTION
**Pull request by SwayamInSync**
# What does this PR do?

Fixes #2421 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [ ] Did you write any new necessary tests?

## Who can review?
@qgallouedec

### CODE REVIEW

**Review by kashif — APPROVED** [review]

File: trl/trainer/dpo_config.py [hunk 1]
```diff
@@ -141,6 +141,8 @@ class DPOConfig(TrainingArguments):
             for saving memory and speeding up training by not computing the logits for all tokens, especially in scenarios
             when working with very long prompts where labels are -ignored (-100).
             [Read more](https://huggingface.co/docs/transformers/main/model_doc/llama#transformers.LlamaForCausalLM)
+        precompute_ref_batch_size (`int`, *optional*, defaults to `None`):
```
Comment (qgallouedec) [inline trl/trainer/dpo_config.py:144]: Can you:

- move this just after `precompute_ref_log_probs`
- mention that if left empty (`None`), `per_device_train_batch_size` is used

**Comment by kashif** [comment]
also if you can kindly add this config to the tests where we test with `pre_compute` flag, then we can also have it tested

**Comment by qgallouedec** [comment]
Good point @kashif 

```python
    def test_precompute_ref_batch_size(self):
        with tempfile.TemporaryDirectory() as tmp_dir:
            training_args = DPOConfig(
                output_dir=tmp_dir,
                per_device_train_batch_size=2,
                precompute_ref_log_probs=True,
                precompute_ref_batch_size=4,
                report_to="none",
            )

            dummy_dataset = load_dataset("trl-internal-testing/zen", "standard_preference")
            model = self.model
            ref_model = self.ref_model
            tokenizer = self.tokenizer

            trainer = DPOTrainer(
                model=model,
                ref_model=ref_model,
                args=training_args,
                processing_class=tokenizer,
                train_dataset=dummy_dataset["train"],
                eval_dataset=dummy_dataset["test"],
            )

            previous_trainable_params = {n: param.clone() for n, param in trainer.model.named_parameters()}

            trainer.train()

            self.assertIsNotNone(trainer.state.log_history[-1]["train_loss"])

            # check the params have changed
            for n, param in previous_trainable_params.items():
                new_param = trainer.model.get_parameter(n)
                # check the params have changed - ignore 0 biases
                if param.sum() != 0:
                    self.assertFalse(torch.allclose(param, new_param, rtol=1e-12, atol=1e-12))
```

File: trl/trainer/dpo_config.py [hunk 2]
```diff
@@ -188,6 +190,7 @@ class DPOConfig(TrainingArguments):
     rpo_alpha: Optional[float] = None
     discopop_tau: float = 0.05
     use_num_logits_to_keep: bool = False
+    precompute_ref_batch_size: Optional[int] = None
```
Comment (qgallouedec) [inline trl/trainer/dpo_config.py:193]: could you move it up as well?

File: trl/trainer/dpo_config.py [hunk 3]
```diff
@@ -94,6 +94,8 @@ class DPOConfig(TrainingArguments):
         precompute_ref_log_probs (`bool`, *optional*, defaults to `False`):
             Whether to precompute reference model log probabilities for training and evaluation datasets. This is
             useful when training without the reference model to reduce the total GPU memory needed.
+        precompute_ref_batch_size (`int`, *optional*, defaults to `None`):
+            Batch size to use when precomputing reference model log probabilities. This can be set higher than the training batch size to speed up preprocessing. If None, defaults to per_device_train_batch_size for training and per_device_eval_batch_size for evaluation.
```
Comment (qgallouedec) [inline trl/trainer/dpo_config.py:98]: just a few line breaks

```suggestion
        precompute_ref_batch_size (`int`, *optional*, defaults to `None`):
            Batch size to use when precomputing reference model log probabilities. This can be set higher than the
            training batch size to speed up preprocessing. If `None``, defaults to `per_device_train_batch_size` for
            training and `per_device_eval_batch_size` for evaluation.
```

File: trl/trainer/dpo_config.py [hunk 4]
```diff
@@ -94,6 +94,10 @@ class DPOConfig(TrainingArguments):
         precompute_ref_log_probs (`bool`, *optional*, defaults to `False`):
             Whether to precompute reference model log probabilities for training and evaluation datasets. This is
             useful when training without the reference model to reduce the total GPU memory needed.
+        precompute_ref_batch_size (`int`, *optional*, defaults to `None`):
+            Batch size to use when precomputing reference model log probabilities. This can be set higher than the
+            training batch size to speed up preprocessing. If `None``, defaults to `per_device_train_batch_size` for
```
Comment (qgallouedec) [inline trl/trainer/dpo_config.py:99]:
```suggestion
            training batch size to speed up preprocessing. If `None`, defaults to `per_device_train_batch_size` for
```

**Review by qgallouedec — APPROVED** [review]
LGTM, I'll merge once the CI is green. Thanks again @SwayamInSync

File: trl/trainer/dpo_config.py [hunk 5]
```diff
@@ -94,6 +94,10 @@ class DPOConfig(TrainingArguments):
         precompute_ref_log_probs (`bool`, *optional*, defaults to `False`):
             Whether to precompute reference model log probabilities for training and evaluation datasets. This is
             useful when training without the reference model to reduce the total GPU memory needed.
+        precompute_ref_batch_size (`int`, *optional*, defaults to `None`):
```
Comment (qgallouedec) [inline trl/trainer/dpo_config.py:97]:
```suggestion
        precompute_ref_batch_size (`Optional[int]`, *optional*, defaults to `None`):
```
